presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -620,6 +624,12 @@ private static boolean isScheduled(Optional<StageInfo> rootStage)
 
     public boolean setFinalQueryInfo(QueryInfo currentInfo, QueryInfo newInfo)
     {
+        queryInfoState.set(FINAL_INFO_AVAILABLE);
         return finalQueryInfo.compareAndSet(currentInfo, newInfo);
```

Comment:
The value must be set before the state is changed, or a lister can get a null before the set is called.